wrappers/rust/icicle-cuda-runtime/src/memory.rs

```diff
@@ -2,37 +2,67 @@ use crate::bindings::{cudaFree, cudaMalloc, cudaMallocAsync, cudaMemcpy, cudaMem
 use crate::error::{CudaError, CudaResult, CudaResultWrap};
 use crate::stream::CudaStream;
 use std::mem::{size_of, MaybeUninit};
+use std::ops::{Index, IndexMut, Range, RangeFrom, RangeFull, RangeInclusive, RangeTo, RangeToInclusive};
 use std::os::raw::c_void;
-use std::slice;
+use std::slice::from_raw_parts_mut;
 
-/// Fixed-size device-side slice.
-#[derive(Debug)]
-#[repr(C)]
-pub struct DeviceSlice<'a, T>(&'a mut [T]);
+pub enum HostOrDeviceSlice<'a, T> {
+    Host(Vec<T>),
+    Device(&'a mut [T]),
+}
 
-impl<'a, T> DeviceSlice<'a, T> {
+impl<'a, T> HostOrDeviceSlice<'a, T> {
     pub fn len(&self) -> usize {
-        self.0
-            .len()
+        match self {
+            Self::Device(s) => s.len(),
+            Self::Host(v) => v.len(),
+        }
     }
 
     pub fn is_empty(&self) -> bool {
-        self.0
-            .is_empty()
+        match self {
+            Self::Device(s) => s.is_empty(),
+            Self::Host(v) => v.is_empty(),
+        }
     }
 
-    pub fn as_slice(&mut self) -> &mut [T] {
-        self.0
+    pub fn is_on_device(&self) -> bool {
```

Comment:
Do we want to align this with the `are_on_device` in the NTT and MSM configs?

Reply:
If I understand you correctly, this is implemented. On device flags in configs are private and are set depending on on-deviceness of data, see for example https://github.com/ingonyama-zk/icicle/pull/333/files#diff-8ca95f9dad1c6679d35f1eacde6413f991f6a34e7bd2249c12d610a121abf10fR104

Reply:
Oo i just meant the naming of the variables so the example you showed would be 

local_cfg.are_inputs_on_device = input.**are**_on_device();

Reply:
Oh. We can I guess, I just named it like that because "slice" is singular